sys/ztimer64/ztimer64.c

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (C) 2021 Kaspar Schleiser <kaspar@schleiser.de>
+ *               2021 Freie Universität Berlin
+ *               2021 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     sys_ztimer
+ * @{
+ *
+ * @file
+ * @brief       ztimer64 core functionality
+ *
+ * This file contains ztimer64's main API implementation and functionality
+ *
+ * TODO:
+ * - add some kind of tick so base_clock overflows don't get lost
+ * - TEST
+ * - add proper unittests
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+#include <assert.h>
+#include <stdint.h>
+#include <inttypes.h>
+
+#include "ztimer64.h"
+
+#define ZTIMER64_CHECKPOINT_INTERVAL (1LU << 31)
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static int _add_entry_to_list(ztimer64_clock_t *clock, ztimer64_base_t *entry);
+static int _del_entry_from_list(ztimer64_clock_t *clock,
+                                ztimer64_base_t *entry);
+static void _ztimer64_update(ztimer64_clock_t *clock);
+
+/* debug aid */
+void ztimer64_clock_print(const ztimer64_clock_t *clock);
+
+static unsigned _is_set(const ztimer64_t *t)
+{
+    /* no timer can ever trigger at `target==0`, so we use that
+     * as marker */
+    return !!(t->base.target);
+}
+
+unsigned ztimer64_is_set(const ztimer64_t *timer)
+{
+    return _is_set(timer);
+}
+
+void ztimer64_remove(ztimer64_clock_t *clock, ztimer64_t *timer)
+{
+    unsigned state = irq_disable();
+
+    if (_is_set(timer)) {
+        if (_del_entry_from_list(clock, &timer->base) == 0) {
+            _ztimer64_update(clock);
+        }
+    }
+
+    irq_restore(state);
+}
+
+void ztimer64_set_at(ztimer64_clock_t *clock, ztimer64_t *timer, uint64_t val)
+{
+    unsigned state = irq_disable();
+
+    if (_is_set(timer)) {
+        _del_entry_from_list(clock, &timer->base);
+    }
+
+    timer->base.target = val;
```

Comment:
Does it make sense to add an assert above this line to verify the assumption [here](https://github.com/RIOT-OS/RIOT/pull/16928/files#diff-2ef78120eb65c80e3c55ceb3089bb542bed8f3dd45abdf35c52954daa6b5c45aR49-R50)

Reply:
not necessary I think, the timer would be added to the list, but as it's time is <now, it would trigger an immediate interrupt.

Reply:
ah, there might be a race, between setting that timer with target==0, before the interrupt hits, if someone calls "is_set()" in the meantime ...

Reply:
I think then I'd maybe rather do "if target==0 target==now()".

Reply:
Also fine with me :)

Reply:
But then please add something to the doxygen to document the behaviour

Reply:
ok, in "set_at()", we don't have "now()", which is expensive to get. but _ztimer64_update() will be called before releasing interrupts, so I've moved the logic there. and added comments. :)